File: android/src/main/java/com/newlogic/mlkit/MLKitPlugin.java
package com.newlogic.mlkit;

import android.app.Activity;
import android.content.Intent;

import com.getcapacitor.JSObject;
import com.getcapacitor.NativePlugin;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.newlogic.mlkitlib.newlogic.MLKitActivity;
import com.newlogic.mlkitlib.newlogic.config.Config;

import org.json.JSONException;
import org.json.JSONObject;

import timber.log.Timber;

import static com.newlogic.mlkitlib.newlogic.config.Actions.START_MLKIT;

/**
 * Sample capacitor call:
 * ===============================================================================
 * const result = await MLKitPlugin.executeMLKit({
 *         action: 'START_MLKIT',
 *         mode: 'mrz',
 *         mrzFormat: 'MRTD_TD1', // Format MRTD_TD1, but default is MRP
 *         config: {
 *           branding: false, // if true, ID pass branding will appear
 *           background: '#ffc234', // default color gray if empty, will accept hex color values only
 *           font: 'NOTO_SANS_ARABIC',
 *           imageResultType: 'base_64', // default path if empty or not set to base64
 *           label: 'التقط الصورة',
 *           isManualCapture: true // if true, manual capture button will appear
 *         }
 *      }
 *  );
 * ==================================================================================
* */
@NativePlugin(requestCodes = {MLKitPlugin.REQUEST_OP_MLKIT})
public class MLKitPlugin extends Plugin {

    public static final int REQUEST_OP_MLKIT = 1001;

    @PluginMethod
    public void startMLActivity(PluginCall call) {
        String mode = call.getString("mode");
        Timber.d("startMLActivity %s", mode);
        saveCall(call);
        if (mode.equals("mrz")) {
            Intent intent = new Intent(getActivity(), MLKitActivity.class);
            intent.putExtra("mode", mode);
            startActivityForResult(call, intent, REQUEST_OP_MLKIT);
        } else {
            call.error("\"" + mode + "\" is not a recognized mode.");
        }
    }

    @PluginMethod
    public void executeMLKit(PluginCall call) {
        JSONObject config = call.getObject("config");
        String action = call.getString("action");
        String mode = call.getString("mode");
        String mrzFormat = call.getString("mrzFormat");
        saveCall(call);
        if (action.equals(START_MLKIT.getValue())) {
            Timber.d("executeMLKit %s", action);
            Activity activity = getActivity();
            Intent intent = new Intent(activity, MLKitActivity.class);
            try {
                String background = config.getString("background");
                String font = config.getString("font");
                String label = config.getString("label");
                String imageResultType = config.getString("imageResultType");
                boolean branding = config.getBoolean("branding");
                boolean isManualCapture = call.getBoolean("isManualCapture");
                Config readerConfig = new Config(branding, background, font,imageResultType, label, isManualCapture);
                intent.putExtra(MLKitActivity.MODE, mode);
                intent.putExtra(MLKitActivity.MRZ_FORMAT, mrzFormat);
                intent.putExtra(MLKitActivity.CONFIG, readerConfig);
            } catch (JSONException e) {
                e.printStackTrace();
            }
            startActivityForResult(call, intent, REQUEST_OP_MLKIT);
        } else {
            call.error("\"" + action + "\" is not a recognized action.");
        }
    }

    @Override
    protected void handleOnActivityResult(int requestCode, int resultCode, Intent data) {
        super.handleOnActivityResult(requestCode, resultCode, data);
        PluginCall savedCall = getSavedCall();
        if (savedCall == null) {
            return;
        }
        if (requestCode == REQUEST_OP_MLKIT) {
            Timber.d("Plugin post MLKit Activity resultCode %d", resultCode);
            if (resultCode == Activity.RESULT_OK) {
                String returnedResult = data.getStringExtra(MLKitActivity.MLKIT_RESULT);
                Timber.d("Plugin post MLKit Activity result %s", returnedResult);
                try {
                    JSONObject result = new JSONObject(returnedResult);
                    JSObject ret = new JSObject();
                    ret.put(MLKitActivity.MLKIT_RESULT, result);
                    savedCall.success(ret);
                } catch (JSONException e) {
                    e.printStackTrace();
                }
            } else if (resultCode == Activity.RESULT_CANCELED) {
                Timber.d("Plugin post MLKit Activity. RESULT CANCELLED");
                savedCall.error("Scanning Cancelled.");
            } else {
                savedCall.error("Scanning Failed.");
            }
        } else {
            savedCall.error("Unknown Request Code!");
        }
    }
}

File: android/src/main/java/com/newlogic/mlkit/MLKitPlugin.kt
package com.newlogic.mlkit

import android.app.Activity
import android.content.Intent
import com.getcapacitor.*
import com.google.gson.Gson
import com.newlogic.mlkitlib.newlogic.SmartScannerActivity
import com.newlogic.mlkitlib.newlogic.platform.ScannerOptions
import org.json.JSONException
import org.json.JSONObject
import timber.log.Timber

@NativePlugin(requestCodes = [MLKitPlugin.REQUEST_OP_SCANNER])
class MLKitPlugin : Plugin() {

    companion object {
        const val REQUEST_OP_SCANNER = 1001
    }

    @PluginMethod
    fun executeScanner(call: PluginCall) {
        val action = call.getString("action")
        val options: JSONObject = call.getObject("options")
        saveCall(call)
        if (action == "START_SCANNER") {
            Timber.d("executeScanner %s", action)
            val activity: Activity = activity
            val intent = Intent(activity, SmartScannerActivity::class.java)
            val scannerOptions = Gson().fromJson(options.toString(), ScannerOptions::class.java)
            intent.putExtra(SmartScannerActivity.SCANNER_OPTIONS, scannerOptions)
            startActivityForResult(call, intent, REQUEST_OP_SCANNER)
        } else {
            call.error("\"$action\" is not a recognized action.")
        }
    }

    override fun handleOnActivityResult(requestCode: Int, resultCode: Int, data: Intent) {
        super.handleOnActivityResult(requestCode, resultCode, data)
        val savedCall = savedCall ?: return
        if (requestCode == REQUEST_OP_SCANNER) {
            Timber.d("Plugin post MLKit Activity resultCode %d", resultCode)
            if (resultCode == Activity.RESULT_OK) {
                val returnedResult = data.getStringExtra(SmartScannerActivity.MLKIT_RESULT)
                Timber.d("Plugin post MLKit Activity result %s", returnedResult)
                try {
                    val result = JSONObject(returnedResult)
                    val ret = JSObject()
                    ret.put(SmartScannerActivity.MLKIT_RESULT, result)
                    savedCall.success(ret)
                } catch (e: JSONException) {
                    e.printStackTrace()
                }
            } else if (resultCode == Activity.RESULT_CANCELED) {
                Timber.d("Plugin post MLKit Activity. RESULT CANCELLED")
                savedCall.error("Scanning Cancelled.")
            } else {
                savedCall.error("Scanning Failed.")
            }
        } else {
            savedCall.error("Unknown Request Code!")
        }
    }
}
